pandas/_libs/tslibs/src/datetime/pd_datetime.h

```diff
@@ -107,6 +111,12 @@ static PandasDateTime_CAPI *PandasDateTimeAPI = NULL;
                                             (base))
 #define make_iso_8601_timedelta(tds, outstr, outlen)                           \
   PandasDateTimeAPI->make_iso_8601_timedelta((tds), (outstr), (outlen))
+#define hash_datetime_from_struct(dts)                                         \
+  PandasDateTimeAPI->hash_datetime_from_struct((dts))
+#define np_datetime64_object_hash(dts)                                         \
+  PandasDateTimeAPI->np_datetime64_object_hash((key))
+#define tuple_update_uhash(acc, lane)                                          \
```

Comment:
OK need to be a bit careful here. Remember that C has no concept of namespaces - if you define `tuple_update_hash` here and also in `khash.h`, then whichever code being insert will depend upon whichever one of those files was last included, which can be very confusing.

With the capsule, the function that you ultimately include as part of the `PandasDateTime_CAPI` struct should be defined as `static` within the `pd_datetime.c` module, and the macro which resolves to that should have a unique name

Reply:
tuple_update_hash is accessed in khash_python.h, is only defined in np_datetime.c AFAICT

Reply:
Header files are loosely just copied into whatever they are included into. So you are doing `#include khash_python.h` with one declaration of `tuple_update_uhash` and then doing `#include pd_datetime.h` which has a different notion of what `tuple_update_uhash` is, your target ultimately gets two different ideas for what `tuple_update_uhash` means. That's basically what the current build failures are telling you

Reply:
How do they have a different notion of what tuple_update_uhash is?  It's only defined in one place.

Opened #51951 to revert the capsuling.

Reply:
You have a declaration for `tuple_update_uhash` is khash_python.h then try to define it via a macro in `pd_datetime.h`

I understand it is frustrating but the implementation here is not correct. Have you checked out any books on C programming? K&R is a nice book for a quick intro - might be worth reading through that or another book before coming back to this

Reply:
> You have a declaration for tuple_update_uhash in khash_python.h 

I'm grepping for "tuple_update_uhash" in khash_python.h and the only place I see it is used inside tupleobject_hash.  Where is this declaration you're referring to?

Reply:
Ah sorry not khash_python looks like np_datetime

Reply:
OK, so if i remove it from np_datetime.h i get build failure locally `error: implicit declaration of function 'tuple_update_uhash' is invalid in C99`

Reply:
That's right. The way the includes happen and the way you have things structured now there you are either providing conflicting declarations for `tuple_update_uhash` for files that aren't `pd_datetime.c` OR you are providing no declaration for `pd_datetime.c`. You would probably be much better served if you had your hashing functions defined as `static` functions within the capsule rather than your current structure.

I _really_ think you'd be better off getting an intro to C book and learning a bit more about what headers/modules do with respect to the different compilation phases. Of course always good to get hands-on experience, but doing some of the basics first before trying to dive in would probably make the process easier

Reply:
Will I appreciate your encouragement but I have a long reading list and dayjob responsibilities.  Can you just tell me how to fix the [expletive deleted] thing that was working a week ago?

Reply:
Yes. Move your hashing functions to the capsule as static functions